File: leetcode/27.remove-element/ts/removeelement.ts
const removeElement = (nums: number[], val: number): number => {
    return removeElementOptimized(nums, val);
};

const removeElementOptimized = (nums: number[], val: number): number => {
    let s = 0

    for (let i = 0; i < nums.length; i++) {
        if (nums[i] === val) continue

        swap(nums, i, s);
        s++
    }

    return s
};

const swap = (nums: number[], a: number, b: number) => {
    const temp = nums[a]
    nums[a] = nums[b]
    nums[b] = temp
}

export default removeElement;

File: leetcode/27.remove-element/cpp/removeelement.cpp
#include <vector>

class Solution
{
public:
    int removeElement(std::vector<int> &nums, int val)
    {
        return removeElementOptimized(nums, val);
    }

private:
    // N, 1
    int removeElementOptimized(std::vector<int> &nums, int val)
    {
        int retainedCount = 0;

        for (int i = 0; i < nums.size(); ++i)
        {
            if (nums[i] == val)
                continue;

            swap(nums, i, retainedCount);
            retainedCount++;
        }

        return retainedCount;
    }

    void swap(std::vector<int> &nums, int a, int b)
    {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
};
